internal/charmstore/store.go

```diff
@@ -505,6 +505,114 @@ func (s *Store) UpdateBaseEntity(url *charm.Reference, update interface{}) error
 	return nil
 }
 
+// Promulgate sets the base entity of url to be promulgated, and unsets
+// promulgated on any other base entity for charms with the same name. It
+// also calculates the next promulgated URL for the charms owned by the
+// new owner and sets those charms.
+//
+// Note: This code is known to have some unfortunate (but not dangerous)
+// race conditions. It is possible that if one or more promulgations
+// happens concurrently for the same entity name then it could result in
+// more than one base entity being promulgated. If this happens then
+// uploads to either user will get promulgated names, these names will
+// never clash. This situation is easily remidied by setting the
+// promulgated user for this charm again, even to one of the ones that is
+// already promulgated. It can also result in the latest promulgated
+// revision of the charm not being one created by the promulgated user,
+// this will be remedied when a new charm is uploaded by the promulgated
+// user. As promulgation is a rare operation, it is considered that the
+// chances this will happen are slim.
+func (s *Store) Promulgate(url *charm.Reference) error {
+	entity, err := s.FindEntity(url)
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	bURL := baseURL(entity.URL)
+
+	// Clear the promulgated flag on any Base entities with the same name,
+	// but aren't the one that is being set.
+	_, err = s.DB.BaseEntities().UpdateAll(
+		bson.D{
+			{"_id", bson.D{{"$ne", bURL}}},
+			{"name", bURL.Name},
+			{"promulgated", 1},
```

Comment:
I guess we *could* use mongodoc.IntBool here and below,
to make the association clear to the reader.